src/js/github.js

```diff
@@ -0,0 +1,123 @@
+
+	return "https://github.com/login/oauth/authorize"
+		+ "?client_id=" + OAUTH_CLIENT_ID
+		+ "&scope=gist"
+		+ "&state=" + randomState
+		+ "&allow_signup=true";
+};
+
+github.signOut = function() {
+	storage_remove("oauth_access_token");
+};
+
+github.isSignedIn = function() {
+	var token = storage_get("oauth_access_token");
+	return typeof token === "string";
+};
+
+github.load = function(id, done) { 
+	var githubURL = "https://api.github.com/gists/"+id;
+
+	var githubHTTPClient = new XMLHttpRequest();
+	githubHTTPClient.open("GET", githubURL);
+	githubHTTPClient.onreadystatechange = function() {
+		if (githubHTTPClient.readyState!=4) {
+			return;
+		} else if (githubHTTPClient.responseText==="") {
+			done(null, "GitHub request returned nothing.  A connection fault, maybe?");
+			return;
+		}
+
+		var limit = window.parseInt(githubHTTPClient.getResponseHeader("x-ratelimit-limit"));
+		var used = window.parseInt(githubHTTPClient.getResponseHeader("x-ratelimit-used"));
+		var reset = new Date(1000 * window.parseInt(githubHTTPClient.getResponseHeader("x-ratelimit-reset")));
+		console.log("Rate limit used " + used + "/" + limit + " (resets " + reset.toISOString() + ")");
+
+		var result = JSON.parse(githubHTTPClient.responseText);
+		if (githubHTTPClient.status===403) {
+			if (!github.isSignedIn() && (result.message.indexOf("rate limit") !== -1)) {
+				done(null, "Exceeded GitHub rate limits. Try signing in from the editor.");
+			} else {
+				done(null, result.message);
+			}
+		} else if (githubHTTPClient.status===401) {
+			github.signOut();
+			done(null, "Authorization check failed.  Try reloading or signing back in from the editor.");
+		} else if (githubHTTPClient.status>=500) {
+			done(null, "HTTP Error "+ githubHTTPClient.status + " - " + githubHTTPClient.statusText + ".");
+		} else if (githubHTTPClient.status!==200 && githubHTTPClient.status!==201) {
+			done(null, "HTTP Error "+ githubHTTPClient.status + " - " + githubHTTPClient.statusText);
+		} else {
+			var result = JSON.parse(githubHTTPClient.responseText);
+			var code=result["files"]["script.txt"]["content"];
+			done(code, null);
+		}
+	}
+
+	if (github.isSignedIn()) {
+		var oauthAccessToken = storage_get("oauth_access_token");
+		githubHTTPClient.setRequestHeader("Authorization", "Token "+oauthAccessToken);
+	}
+	githubHTTPClient.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
+	githubHTTPClient.send();
+};
+
+github.save = function(title, code, done) {
+	var oauthAccessToken = storage_get("oauth_access_token");
+	if (typeof oauthAccessToken !== "string") {
+		printUnauthorized();
+		return;
+	}
+
+	var gistToCreate = {
+		"description" : title,
+		"public" : true,
+		"files": {
+			"readme.txt" : {
+				"content": "Play this game by pasting the script in http://www.puzzlescript.net/editor.html"
+			},
+			"script.txt" : {
+				"content": code
+			}
+		}
+	};
+
+	var githubURL = "https://api.github.com/gists";
```

Comment:
nit: Maybe worth defining this at the top, similar to `OAUTH_CLIENT_ID`.

Reply:
nah it's ok